src/mlpack/methods/ann/rbm/rbm.hpp

```diff
@@ -0,0 +1,458 @@
+   */
+  template<typename Policy = PolicyType>
+  typename std::enable_if<std::is_same<Policy, SpikeSlabRBM>::value, void>::type
+  SampleSpike(DataType&& spikeMean, DataType&& spike);
+
+  /**
+   * The function calculates the mean of Normal distribution of P(s|v, h),
+   * where the mean is given by:
+   * $h_i*\alpha^{-1}*W_i^T*v$
+   * 
+   * @param visible The visible layer neurons.
+   * @param spike The spike variables from hidden layer.
+   * @param slabMean The mean of the Normal distribution of slab neurons.
+   */
+  template<typename Policy = PolicyType>
+  typename std::enable_if<std::is_same<Policy, SpikeSlabRBM>::value, void>::type
+  SlabMean(DataType&& visible, DataType&& spike, DataType&& slabMean);
+
+  /**
+   * The function samples from the Normal distribution of P(s|v, h),
+   * where the mean is given by:
+   * $h_i*\alpha^{-1}*W_i^T*v$
+   * and variance is given by:
+   * $\alpha^{-1}$
+   *
+   * @param slabMean Mean of the Normal distribution of the slab neurons.
+   * @param slab Sampled slab variable from the Normal distribution.
+   */
+  template<typename Policy = PolicyType>
+  typename std::enable_if<std::is_same<Policy, SpikeSlabRBM>::value, void>::type
+  SampleSlab(DataType&& slabMean, DataType&& slab);
+
+  /**
+   * This function does the k-step Gibbs Sampling.
+   *
+   * @param input Input to the Gibbs function.
+   * @param output Used for storing the negative sample.
+   * @param steps Number of Gibbs Sampling steps taken.
+   */
+  void Gibbs(arma::Mat<ElemType>&& input,
+             arma::Mat<ElemType>&& output,
+             const size_t steps = SIZE_MAX);
+
+  /**
+   * Calculates the gradients for the RBM network.
+   *
+   * @param parameters The current parameters of the network.
+   * @param i Index of the data point.
+   * @param gradient Variable to store the present gradient.
+   * @param batchSize Variable to store the present number of inputs.
+   */
+  void Gradient(const arma::Mat<ElemType>& parameters,
+                const size_t i,
+                arma::Mat<ElemType>& gradient,
+                const size_t batchSize);
+
+  /**
+   * Shuffle the order of function visitation. This may be called by the
+   * optimizer.
+   */
+  void Shuffle();
+
+  //! Return the number of separable functions (the number of predictor points).
+  size_t NumFunctions() const { return numFunctions; }
+
+  //! Return the number of steps of Gibbs Sampling.
+  size_t NumSteps() const { return numSteps; }
+
+  //! Return the parameters of the network.
+  const arma::Mat<ElemType>& Parameters() const { return parameter; }
+  //! Modify the parameters of the network.
+  arma::Mat<ElemType>& Parameters() { return parameter; }
+
+  //! Return the weights of the network.
+  const DataType& Weight() const { return weight; }
+  //! Modify the weights of the network.
+  DataType& Weight() { return weight; }
+  //! Get the weight of the network.
+  arma::cube const& WeightCube() const { return weightCube; }
```

Comment:
Do you think we can remove the function? Not sure it's necessary, we could do the transformation based on the ``Weight`` function as well.

Reply:
This is necessary for the `SSRBM` class, as it uses cubic parameters (unlike matrix parameters).

Reply:
Sure, but it does use the internal ``weightCube`` parameter and not the ``WeightCube()``method right?